File: src/LorAuto.SDK.Python/LorAuto/Plugin/StrategyPlugin.py
from abc import abstractmethod
from typing import Tuple, Optional

from LorAuto.Card.Model.EGameCardType import EGameCardType
from LorAuto.Card.Model.InGameCard import InGameCard
from LorAuto.Client.CardTargetSelector import CardTargetSelector
from LorAuto.Client.Model.BoardCards import BoardCards
from LorAuto.Client.Model.EGamePlayAction import EGamePlayAction
from LorAuto.Client.Model.EGameState import EGameState
from LorAuto.Client.Model.GameBoardData import GameBoardData
from LorAuto.Plugin.PluginBase import PluginBase


class StrategyPlugin(PluginBase):
    """
    Base class for implementing strategies.
    """

    def GetPlayableHandCards(self, board_cards: BoardCards, mana: int, spell_mana: int) -> list[InGameCard]:
        """
        Gets the list of playable hand cards based on the current board state and available resources.

        Args:
            board_cards (BoardCards): The board cards containing the player's hand cards.
            mana (int): The available mana.
            spell_mana (int): The available spell mana.

        Returns:
            List[InGameCard]: A list of playable hand cards.
        """
        pass

    @abstractmethod
    def Mulligan(self, mulligan_cards: list[InGameCard]) -> list[InGameCard]:
        """
        Performs the mulligan phase, selecting cards to replace.

        Args:
            mulligan_cards (List[InGameCard]): The cards available for mulligan.

        Returns:
            List[InGameCard]: The list of cards to replace.
        """
        pass

    @abstractmethod
    def PlayHandCard(self, board_data: GameBoardData, game_state: EGameState, mana: int,
                       spell_mana: int) -> Optional[Tuple[InGameCard, Optional[CardTargetSelector]]]:
        """
        Plays a hand card from the player's hand.

        Args:
            board_data (GameBoardData): The current game board data.
            game_state (EGameState): The current game state.
            mana (int): The available mana.
            spell_mana (int): The available spell mana.

        Returns:
            Optional[Tuple[InGameCard, Optional[CardTargetSelector]]]: A tuple containing the played hand card and
                its target selector (if applicable).
        """
        pass

    @abstractmethod
    def Block(self, board_data: GameBoardData,
              spells_to_use: list[CardTargetSelector]) -> dict[InGameCard, InGameCard]:
        """
        Blocks incoming attacks from the opponent's board cards.

        Args:
            board_data (GameBoardData): The current game board data.
            spells_to_use (List[CardTargetSelector]): Output parameter for the list of spell cards to use for blocking.

        Returns:
            Dict[InGameCard, InGameCard]: A dictionary mapping player's own board cards to the opponent's board cards
            to block them.
        """
        pass

    @abstractmethod
    def RespondToOpponentAction(self, board_data: GameBoardData, game_state: EGameState,
                                   mana: int, spell_mana: int) -> EGamePlayAction:
        """
        Responds to an opponent's action during the game.

        Args:
            board_data (GameBoardData): The current game board data.
            game_state (EGameState): The current game state.
            mana (int): The available mana.
            spell_mana (int): The available spell mana.

        Returns:
            EGamePlayAction: The action to perform in response to the opponent's action.
        """
        pass

    @abstractmethod
    def AttackTokenUsage(self, board_data: GameBoardData, mana: int, spell_mana: int) -> EGamePlayAction:
        """
        Determines the action to take for using attack tokens.

        Args:
            board_data (GameBoardData): The current game board data.
            mana (int): The available mana.
            spell_mana (int): The available spell mana.

        Returns:
            EGamePlayAction: The action to take for using attack tokens.
        """
        pass

    @abstractmethod
    def Attack(self, board_data: GameBoardData, player_board_cards: list[InGameCard]) -> list[InGameCard]:
        """
        Performs the attack action on the opponent's board.

        Args:
            board_data (GameBoardData): The current game board data.
            player_board_cards (List[InGameCard]): The cards on the player's board.

        Returns:
            List[InGameCard]: A list of cards to attack with.
        """
        pass

File: src/LorAuto/Plugin/Types/StrategyPlugin.cs
﻿using LorAuto.Card;
using LorAuto.Card.Model;
using LorAuto.Client.Model;

namespace LorAuto.Plugin.Types;

// TODO: What if bot want to summon but board is full (6 card in board)
// TODO: Maybe add `AcceptAttack` and `AcceptBlock` to make sure cards are moved, as card could be stunned

/// <summary>
/// Base class for implementing strategies.
/// </summary>
public abstract class StrategyPlugin : PluginBase
{
    /// <summary>
    /// Gets the list of playable hand cards based on the current board state and available resources.
    /// </summary>
    /// <param name="boardData">The current game board data.</param>
    /// <returns>A list of playable hand cards.</returns>
    public virtual List<InGameCard> GetPlayableHandCards(GameBoardData boardData)
    {
        return boardData.Cards.CardsHand
            .Where(card =>
                (card.Cost <= boardData.Mana ||
                 (card.Type == EGameCardType.Spell && card.Cost <= boardData.Mana + boardData.SpellMana)) &&
                (boardData.GameState == GameState.MidAttack && card.Type != EGameCardType.Unit))
            .OrderByDescending(card => card.Cost)
            .ToList();
    }

    /// <summary>
    /// Performs the mulligan phase, selecting cards to replace.
    /// </summary>
    /// <param name="mulliganCards">The cards available for mulligan.</param>
    /// <returns>The list of cards to replace.</returns>
    public abstract List<InGameCard> Mulligan(IEnumerable<InGameCard> mulliganCards);

    /// <summary>
    /// Plays a hand card from the player's hand.
    /// </summary>
    /// <param name="boardData">The current game board data.</param>
    /// <param name="gameState">The current game state.</param>
    /// <returns>A tuple containing the played hand card and its target selector (if applicable).</returns>
    public abstract (InGameCard HandCard, CardTargetSelector? Target)? PlayHandCard(
        GameBoardData boardData,
        GameState gameState
    );

    /// <summary>
    /// Blocks incoming attacks from the opponent's board cards.
    /// </summary>
    /// <param name="boardData">The current game board data.</param>
    /// <param name="spellsToUse">Output parameter for the list of spell cards to use for blocking.</param>
    /// <returns>A dictionary mapping player's own board cards to the opponent's board cards to block them.</returns>
    public abstract Dictionary<InGameCard, InGameCard> Block(
        GameBoardData boardData,
        out List<CardTargetSelector>? spellsToUse
    );

    /// <summary>
    /// Responds to an opponent's action during the game.
    /// </summary>
    /// <param name="boardData">The current game board data.</param>
    /// <param name="gameState">The current game state.</param>
    /// <returns>The action to perform in response to the opponent's action.</returns>
    public abstract EGamePlayAction RespondToOpponentAction(
        GameBoardData boardData,
        GameState gameState
    );

    /// <summary>
    /// Determines the action to take for using attack tokens.
    /// </summary>
    /// <param name="boardData">The current game board data.</param>
    /// <returns>The action to take for using attack tokens.</returns>
    public abstract EGamePlayAction AttackTokenUsage(GameBoardData boardData);

    /// <summary>
    /// Performs the attack action on the opponent's board.
    /// </summary>
    /// <param name="boardData">The current game board data.</param>
    /// <param name="playerBoardCards">The cards on the player's board.</param>
    /// <returns>A list of cards to attack with.</returns>
    public abstract List<InGameCard> Attack(GameBoardData boardData, List<InGameCard> playerBoardCards);
}